Platform/OpenNtfsDxe/Data.c

```diff
@@ -0,0 +1,566 @@
+  }
+
+  NonRes = (ATTR_HEADER_NONRES *) AttrStart;
+
+  Runlist = (RUNLIST *) AllocateZeroPool(sizeof (RUNLIST));
+  if (Runlist == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  Runlist->Attr = Attr;
+  Runlist->Unit.FileSystem = Attr->BaseMftRecord->File->FileSystem;
+
+  if ((NonRes->DataRunsOffset > BufferSize) ||
+     (NonRes->DataRunsOffset > NonRes->RealSize) ||
+     (NonRes->RealSize > NonRes->AllocatedSize)) {
+    DEBUG((DEBUG_INFO, "NTFS: Non-Resident Attribute is corrupted.\n"));
+    FreePool (Runlist);
+    return EFI_VOLUME_CORRUPTED;
+  } else {
+    Runlist->NextDataRun = (UINT8 *)NonRes + NonRes->DataRunsOffset;
+  }
+
+  Runlist->NextVcn = NonRes->StartingVCN;
+  Runlist->CurrentLcn = 0;
+
+  Runlist->TargetVcn = NonRes->StartingVCN + DivU64x64Remainder (Offset, mClusterSize, NULL);
+  if (Runlist->TargetVcn < NonRes->StartingVCN) {
+    DEBUG((DEBUG_INFO, "NTFS: Overflow: StartingVCN is too large.\n"));
+    FreePool (Runlist);
+    return EFI_VOLUME_CORRUPTED;
+  }
+
+  if ((Offset + Length) > NonRes->RealSize) {
+    DEBUG((DEBUG_INFO, "NTFS: Read out of range\n"));
+    FreePool (Runlist);
+    return EFI_VOLUME_CORRUPTED;
+  }
+
+  if (((NonRes->Flags & FLAG_COMPRESSED) != 0) && // ((Attr->Flags & NTFS_AF_GPOS) == 0))
+      (NonRes->Type == AT_DATA)) {
+    Status = Decompress (Runlist, Offset, Length, Dest);
+
+    FreePool (Runlist);
+    return Status;
+  }
+
+  while (Runlist->NextVcn <= Runlist->TargetVcn) {
+    Status = ReadRunListElement (Runlist);
+    if (EFI_ERROR(Status)) {
+      FreePool (Runlist);
+      return EFI_VOLUME_CORRUPTED;
+    }
+  }
+
+  // if (Attr->Flags & NTFS_AF_GPOS) {
+  //   OffsetInsideCluster = Offset & (mClusterSize - 1);
+  //
+  //   Sector0 = ((Runlist->TargetVcn - Runlist->CurrentVcn + Runlist->CurrentLcn)
+  //             * mClusterSize + OffsetInsideCluster) / mSectorSize;
+  //
+  //   Sector1 = Sector0 + 1;
+  //   if (Sector1 == ((Runlist->NextVcn - Runlist->CurrentVcn + Runlist->CurrentLcn)
+  //                  * mClusterSize) / mSectorSize) {
+  //     Status = ReadRunListElement (Runlist);
+  //     if (EFI_ERROR(Status)) {
+  //       FreePool (Runlist);
+  //       return EFI_VOLUME_CORRUPTED;
+  //     }
+  //
+  //     Sector1 = Runlist->CurrentLcn * mClusterSize / mSectorSize;
+  //   }
+  //
+  //   *(UINT32 *) Dest = (UINT32) Sector0;
+  //   *(UINT32 *) (Dest + 4) = (UINT32) Sector1;
+  //
+  //   FreePool (Runlist);
+  //   return EFI_SUCCESS;
+  // }
+
+  Status = ReadClusters (
+    Runlist,
+    Offset,
+    Length,
+    Dest
+    );
+
+  FreePool (Runlist);
+  return Status;
+}
+
+STATIC
+UINT64
+ReadField (
+  IN CONST UINT8 *Run,
+  IN INT32       FieldSize,
+  IN BOOLEAN     Signed
+  )
+{
+  UINT64 Value = 0;
+  //
+  // Offset to the starting LCN of the previous element is a signed value.
+  // So we must check the most significant bit.
+  //
+  if (Signed && FieldSize && (Run[FieldSize - 1] & 0x80)) {
+    Value = (UINT64) (-1);
```

Comment:
Then this will be overridden by `CopyMem` below?

Reply:
Only partially and most significant bits will remain the same.